### PR TITLE
[WIP] Add GpuMatND with arbitrary dimension support

### PR DESCRIPTION
**Pull request by czgdp1807**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under OpenCV (BSD) License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

This PR aims to add support for arbitrary dimension support as a first step to support the APIs given in https://github.com/opencv/opencv/issues/16433
As of creating this PR only interface is added and comments on the same is awaited.

```
force_builders_only=docs,Custom
buildworker:Custom=linux-4
build_image:Custom=ubuntu-cuda:16.04
```

### CODE REVIEW

File: modules/core/include/opencv2/core/cuda.hpp [hunk 1]
```diff
@@ -337,6 +337,195 @@ class CV_EXPORTS GpuMat
...
+
+    //! default allocator
+    static Allocator* defaultAllocator();
+    static void setDefaultAllocator(Allocator* allocator);
+
+    //! default constructor
+    explicit GpuMatND(Allocator* allocator = defaultAllocator());
+
+    //! constructs GpuMatND of the specified shape and type
+    GpuMatND(int* shape, int type, Allocator* allocator = defaultAllocator());
```
Comment (czgdp1807) [inline modules/core/include/opencv2/core/cuda.hpp]: We can use variable number of arguments as well here using the APIs in `cstdarg` but in other libraries like TF, the shape is passed as a list, so passing it as an array is better and avoids complications of variable number of arguments.

Comment (alalek) [reply modules/core/include/opencv2/core/cuda.hpp]: > int* shape

This is unsafe API - missing `ndims`.

Consider targeting to "master" branch with C++11 support.
Use `std::array` and or `std::vector<int>` for passing ND array layouts.

File: modules/core/include/opencv2/core/cuda.hpp [hunk 2]
```diff
@@ -337,6 +337,195 @@ class CV_EXPORTS GpuMat
...
+    GpuMatND(int* shape, int type, Scalar s, Allocator* allocator = defaultAllocator());
+
+    //! copy constructor
+    GpuMatND(const GpuMatND& m);
+
+    //! constructor for GpuMatND headers pointing to user-allocated data
+    GpuMatND(int* shape, int type, void* data, size_t step = Mat::AUTO_STEP);
+
+    //! builds GpuMatND from host memory (Blocking call)
+    explicit GpuMatND(InputArray arr, Allocator* allocator = defaultAllocator());
```
Comment (alalek) [inline modules/core/include/opencv2/core/cuda.hpp]: Host memory is `Mat` and not `InputArray` (which is some kind of wrapper over multiple types, but it doesn't store any data).

Comment (czgdp1807) [reply modules/core/include/opencv2/core/cuda.hpp]: Something similar is used in `GpuMat`, so do I need to make changes there as well?

**Comment by czgdp1807** [comment]
> There should be interoperability with 2D `GpuMat`

I will try to keep this in mind while implementing the interface.

File: modules/core/include/opencv2/core/cuda.hpp (hunk 1 above)
Comment (czgdp1807) [reply modules/core/include/opencv2/core/cuda.hpp]: > Consider targeting to "master" branch with C++11 support.

Done.

> Use `std::array` and or `std::vector<int>` for passing ND array layouts.

I have used `std::vector<int>`.

**Comment by czgdp1807** [comment]
@YashasSamaga @alalek Please let me know if any change is to be made in the interface.

**Comment by czgdp1807** [comment]
I hope I can start working on the implementation now. As no more comments are received on the interface. Probably we can minor changes during writing the implementation. Thanks.

**Comment by Llcoolsouder** [comment]
Closes #15897

File: modules/core/include/opencv2/core/cuda.hpp [hunk 3]
```diff
@@ -337,6 +337,195 @@ class CV_EXPORTS_W GpuMat
...
+    //! assignment operators
+    GpuMatND& operator =(const GpuMatND& m);
+
+    //! allocates new GpuMatND data unless the GpuMatND already has specified size and type
+    void create(std::vector<int> shape, int type);
+
+    //! decreases reference counter, deallocate the data when reference counter reaches 0
+    void release();
+
+    //! swaps with other smart pointer
```
Comment (YashasSamaga) [inline modules/core/include/opencv2/core/cuda.hpp]: implementation detail spilled into public API?

File: modules/core/include/opencv2/core/cuda.hpp [hunk 4]
```diff
@@ -337,6 +337,195 @@ class CV_EXPORTS_W GpuMat
...
+    //! template version of the above method
+    template<typename _Tp> _Tp* ptr(int y = 0);
+    template<typename _Tp> const _Tp* ptr(int y = 0) const;
+
+    template <typename _Tp> operator PtrStepSz<_Tp>() const;
+    template <typename _Tp> operator PtrStep<_Tp>() const;
+
+    //! creates alternative GpuMatND header for the same data, with different
+    //! number of channels and/or different number of rows
+    GpuMatND reshape(int* new_shape) const;
```
Comment (YashasSamaga) [inline modules/core/include/opencv2/core/cuda.hpp]: how will you figure out the number of dimensions?

**Comment by YashasSamaga** [comment]
@Llcoolsouder But this won't allocate 3D pitched memory.

@czgdp1807 I am not sure if a generic `GpuMatND` should be implemented or if a proxy object (like the one described in #16433) must be used.

The `GpuMat` interface was for C++98. I think it is better to have a C++11 interface (what's the point of having C++98 interface when OpenCV has already moved to C++11). The only problem would be that the interface would not be compatible with the `GpuMat`'s interface. It's just my opinion.

File: modules/core/include/opencv2/core/cuda.hpp (hunk 4 above)
Comment (czgdp1807) [reply modules/core/include/opencv2/core/cuda.hpp]: Ah! I forgot to replace this with `vector<int> new_shape`. Thanks for pointing out.

**Comment by czgdp1807** [comment]
> I think it is better to have a C++11 interface (what's the point of having C++98 interface when OpenCV has already moved to C++11).

Actually, I am not aware about which parts of the interface of `GpuMat` that are there due to C++98. For example, there is something called `Allocator`. Why don't we use `static` methods to reserve space in the memory and call them in the initializer list of constructor? Please let me know if we can do this and which functions can we remove from `GpuMatND`.

**Comment by czgdp1807** [comment]
Hi,
Any thoughts on the ideas given in https://github.com/opencv/opencv/pull/16666#issuecomment-595111320 or can I proceed with the current interface given in the diff?
Please let me know.
Thanks.

**Comment by nglee** [comment]
Is anyone still working on this? If no one's currently looking into it, I'd like to follow up on what's been discussed so far and work on this.

**Comment by YashasSamaga** [comment]
Bumping. Is there any chance of this being completed soon? I think `GpuMat` is currently the only way to pass GPU memory to `InputArray`/`OutputArray`. This is very limiting. And not all GPU data are 2D images.

**Comment by czgdp1807** [comment]
Hi, 

First of all apologies for not responding. It's been a long time since I updated this PR. If anyone's interested, please feel free to take this over and complete it. 

Apologies again. Stay safe.

**Comment by asmorkalov** [comment]
Replaced with https://github.com/opencv/opencv/pull/19259

**Pull request closed without merging**